### PR TITLE
Added Feedback class and integrated it into the Guess class

### PR DESCRIPTION
**Pull request by MayaCherry1**

### CODE REVIEW

File: lib/feedback.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+class Feedback 
+	attr :secret, :guess
+
+
```
Comment (ssunday) [inline lib/feedback.rb]: extra line

File: lib/feedback.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	def initialize (secret, guess)
+		@secret = secret.clone 
+		@guess = guess.clone
+	end
+
+	def generate 
+		red_pegs
+		white_pegs
+		return {:white => count_white, :red => count_red}
```
Comment (ssunday) [inline lib/feedback.rb]: It feels like the red/white pegs could be pulled out into some setup function. Having it called inside generate feels funky.

How about keeping the functions except the return as is in the function, and create a diff function that returns the counts? Also, return isn't need it

File: lib/game_setup.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
+
+class GameSetup
+
+	attr_reader :colors
+	attr_accessor :secret_code
+
+	def initialize 
+		@colors = ['R' , 'O', 'Y', 'G', 'B', 'P']
```
Comment (ssunday) [inline lib/game_setup.rb]: If this doesn't change, pull it out into a constant

File: lib/guess.rb [hunk 4]
```diff
@@ -19,4 +21,8 @@ def check_colors (colors)
 		@code.all? { |c| colors.include?(c) }
 	end
 
+	def set_feedback(secret)
+		@feedback = Feedback.new(secret,@code).generate
```
Comment (ssunday) [inline lib/guess.rb]: spaces between arguments for easier readability

File: lib/guess_loop.rb [hunk 5]
```diff
@@ -3,37 +3,51 @@
...
 
 	def fix_input (input)
 		input.upcase.tr(" ","").chars
 	end
 
-	def print_guesses
-		puts "\n\n"
-		past_guesses.each { |g| puts g.code.join " "}
+	def print_guesses (colors)
+		if (!@valid)
```
Comment (ssunday) [inline lib/guess_loop.rb]: unless @valid

File: spec/game_setup_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+	describe '#colors' do
+		it 'returns the colors array' do
+			expect(game.colors).to eq (colors)
+		end
+	end
+
+	describe '#secret_code' do
+		it 'returns an empty array when initialized' do
+			expect(game.secret_code.class).to eq (Array)
```
Comment (ssunday) [inline spec/game_setup_spec.rb]: `expect(game.secret_code).to be_empty` actually tests according to the spec...this just tests if it's an array, not that it's empty

File: lib/guess_loop.rb [hunk 7]
```diff
@@ -3,37 +3,51 @@
...
+			puts "INVALID INPUT: Please enter the first letter of the colors you wish to guess\n"
+		end
+		puts "Available Colors: #{colors.join(", ")}\n\n"
+		past_guesses.each { |g| puts "#{g.id}: #{g.code.join(" ")}\t red: #{g.feedback[:red]}  white: #{g.feedback[:white]}"}
+
 	end
 
-	def loop_guesses(colors)
+	def loop_guesses(colors,secret)
 		while (@past_guesses.size < 10)
```
Comment (soft-circles) [inline lib/guess_loop.rb]: Try to eliminate magic numbers where you can. You might want to try extracting the 10 here into a constant like `MAXIMUM_GUESS_COUNT`

File: lib/feedback.rb (hunk 2 above)
Comment (soft-circles) [reply lib/feedback.rb]: Agreed

File: lib/end_game.rb [hunk 8]
```diff
@@ -1,20 +1,21 @@
...
-	def print_msg(has_won, secret, num_guesses)
-		if has_won
-			win_game(secret,num_guesses)
-		else
-			lose_game(secret)
-		end
+	def play_again?
+		dialog.play_again
+		again = input.play_again
+		return again.eql?('y')
```
Comment (ssunday) [inline lib/end_game.rb]: return isn't needed

File: lib/game_loop.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+	def initialize(game_state, dialog, input)
+		@valid = true
+		@game_state = game_state
+		@dialog = dialog
+		@input = input
+	end
+
+	def play
+		while (@game_state.remaining_guesses > 0)
```
Comment (ssunday) [inline lib/game_loop.rb:15]: `@game_state.remaining_guesse.positive?`

File: lib/game_loop.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	def check_validity(guess)
+		@valid = guess.valid?
+	end
+
+	def save_guess_if_valid(guess)
+		@game_state.past_guesses.save_guess(guess) if @valid
+	end
+
+	def check_win(guess)
+	 	@game_state.has_won = true if guess.is_win?
```
Comment (ssunday) [inline lib/game_loop.rb]: @game_state.has_won = guess.is_win? be equivalent? if it's false, it's false, if it's true, then it's true anyway

File: spec/game_loop_spec.rb [hunk 11]
```diff
@@ -1,85 +1,50 @@
...
-dialog.welcome
-
-dialog = Dialog.new(output: ->(x) { HTTP.send(x) })
-dialog.welcome
-
-Dialog.new()
+	# describe '#format_input' do
+	# 	let (:output) {['R', 'O', 'Y', 'G']}
+	# 	context('returns the input as an array of uppercase characters') do
+	# 		it 'when the input has spaces' do
```
Comment (ssunday) [inline spec/game_loop_spec.rb]: `xit` marks the test as pending which is usually better than comments if you plan on bringing the tests back at some point

File: lib/game_loop.rb (hunk 9 above)
Comment (MayaCherry1) [reply lib/game_loop.rb:15]: For some reason when I try .positive?, it works when I run the program but fails with my tests. Maybe the version of rspec that I have is out of date?

**Pull request merged**